ext/pdo_firebird/firebird_driver.c

```diff
@@ -881,9 +884,10 @@ static bool firebird_handle_set_attribute(pdo_dbh_t *dbh, zend_long attr, zval *
 					return false;
 				}
 				if (H->date_format) {
-					efree(H->date_format);
+					pefree(H->date_format, dbh->is_persistent);
+					H->date_format = NULL;
 				}
-				spprintf(&H->date_format, 0, "%s", ZSTR_VAL(str));
+				H->date_format = pestrdup(ZSTR_VAL(str), dbh->is_persistent);
```

Comment:
Since we know the length, you should use `pestrndup` instead and pass in `ZSTR_LEN(str)` as well (same below).